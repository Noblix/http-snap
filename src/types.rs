use crate::snapshot_types::Snapshot;
use reqwest::header::HeaderMap;
use serde::ser::{SerializeMap, SerializeSeq, Serializer};
use serde::Serialize;

#[derive(Debug)]
pub struct HttpFile {
    pub options: SnapOptions,
    pub verb: HttpVerb,
    pub url: String,
    pub headers: Vec<Header>,
    pub body: Json,
    pub snapshot: Snapshot,
}

#[derive(Debug)]
pub struct SnapResponse {
    pub options: SnapOptions,
    pub status: u16,
    pub headers: HeaderMap,
    pub body: Json
}

#[derive(Debug, Clone)]
pub struct SnapOptions {
    pub include_headers: bool,
}

#[derive(Debug, Eq, PartialEq)]
pub enum HttpVerb {
    GET,
    DELETE,
    PATCH,
    POST,
    PUT,
}

#[derive(Debug)]
pub struct Header {
    pub name: String,
    pub value: String,
}

#[derive(Debug)]
pub struct Json {
    pub element: Element,
}

impl Serialize for Json {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        return self.element.serialize(serializer);
    }
}

#[derive(Debug)]
pub struct Element {
    pub value: Value,
}

impl Serialize for Element {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        return self.value.serialize(serializer);
    }
}

impl Element {
    pub fn to_insertion_string(&self, nested: bool) -> String {
        return self.value.to_insertion_string(nested);
    }
}

#[derive(Debug)]
pub enum Value {
    Object(Object),
    Array(Array),
    String(String),
    Number(Number),
    Boolean(bool),
    Null(),
}

impl Serialize for Value {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match self {
            Value::Object(val) => val.serialize(serializer),
            Value::Array(val) => val.serialize(serializer),
            Value::String(val) => serializer.serialize_str(val),
            Value::Number(val) => val.serialize(serializer),
            Value::Boolean(val) => serializer.serialize_bool(val.clone()),
            Value::Null() => serializer.serialize_none(),
        }
    }
}

impl Value {
    pub fn to_insertion_string(&self, nested: bool) -> String {
        match self {
            Value::Object(val) => val.to_insertion_string(),
            Value::Array(val) => val.to_insertion_string(),
            Value::String(val) => {
                if nested {
                    ["\"", val, "\""].concat()
                } else {
                    val.clone()
                }
            }
            Value::Number(val) => val.to_insertion_string(),
            Value::Boolean(val) => val.to_string(),
            Value::Null() => "null".to_string(),
        }
    }
}

#[derive(Debug)]
pub struct Object {
    pub members: Vec<Member>,
}

impl Serialize for Object {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut map = serializer.serialize_map(Some(self.members.len()))?;
        for member in &self.members {
            map.serialize_entry(&member.key, &member.value)?;
        }
        return map.end();
    }
}

impl Object {
    pub fn to_insertion_string(&self) -> String {
        let mut result = "{".to_string();
        let mut member_strings = Vec::new();
        for member in &self.members {
            member_strings.push(member.to_insertion_string(true))
        }
        result += &member_strings.join("\n,");
        result += "}";
        return result;
    }
}

#[derive(Debug)]
pub struct Member {
    pub key: String,
    pub value: Element,
}

impl Member {
    pub fn to_insertion_string(&self, nested: bool) -> String {
        return [
            "\"",
            &self.key,
            "\": ",
            &self.value.to_insertion_string(nested),
        ]
        .concat();
    }
}

#[derive(Debug)]
pub struct Array {
    pub elements: Vec<Element>,
}

impl Serialize for Array {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut seq = serializer.serialize_seq(Some(self.elements.len()))?;
        for element in &self.elements {
            seq.serialize_element(element)?;
        }
        return seq.end();
    }
}

impl Array {
    pub fn to_insertion_string(&self) -> String {
        let mut result = "[".to_string();
        let mut element_strings = Vec::new();
        for element in &self.elements {
            element_strings.push(element.to_insertion_string(true))
        }
        result += &element_strings.join("\n,");
        result += "]";
        return result;
    }
}

#[derive(Debug)]
pub enum Number {
    Int(i64),
    Fraction(f64),
    Exponent(String),
}

impl Serialize for Number {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match self {
            Number::Int(val) => serializer.serialize_i64(val.clone()),
            Number::Fraction(val) => serializer.serialize_f64(val.clone()),
            Number::Exponent(val) => serializer.serialize_f64(val.parse::<f64>().unwrap()),
        }
    }
}

impl Number {
    pub fn to_insertion_string(&self) -> String {
        match self {
            Number::Int(val) => val.to_string(),
            Number::Fraction(val) => val.to_string(),
            Number::Exponent(val) => val.clone(),
        }
    }
}
