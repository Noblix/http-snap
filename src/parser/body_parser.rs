use crate::parser::variable_parser::{variable_name_parser, variable_name_string_parser};
use crate::types::*;
use chumsky::error::Simple;
use chumsky::prelude::*;
use chumsky::text::{int, whitespace};
use chumsky::Parser;

// Based on https://www.json.org/json-en.html
pub(crate) fn body_parser() -> impl Parser<char, Json, Error = Simple<char>> {
    let no_body = empty().map(|_| Json {
        element: Element {
            value: Value::Null(),
        },
    });

    return json_parser().or(no_body);
}

fn json_parser() -> impl Parser<char, Json, Error = Simple<char>> {
    return element_parser().map(|element| Json { element });
}

pub(crate) fn value_parser(
    element_parser: impl Parser<char, Element, Error = Simple<char>> + Clone,
) -> impl Parser<char, Value, Error = Simple<char>> {
    let variable = variable_name_string_parser().map(|name| Value::VariableReference(name));
    let boolean = (just("true").map(|_| Value::Boolean(true)))
        .or(just("false").map(|_| Value::Boolean(false)));
    let null = just("null").map(|_| Value::Null());

    return variable
        .or(object_parser(element_parser.clone()))
        .or(array_parser(element_parser.clone()))
        .or(string_value_parser())
        .or(number_value_parser())
        .or(boolean)
        .or(null);
}

fn object_parser(
    element_parser: impl Parser<char, Element, Error = Simple<char>> + Clone,
) -> impl Parser<char, Value, Error = Simple<char>> {
    let empty = whitespace().delimited_by(just("{"), just("}")).map(|_| {
        Value::Object(Object {
            members: Vec::new(),
        })
    });

    let members = members_parser(element_parser)
        .delimited_by(just("{"), just("}"))
        .map(|members| Value::Object(Object { members }));

    return empty.or(members);
}

fn members_parser(
    element_parser: impl Parser<char, Element, Error = Simple<char>> + Clone,
) -> impl Parser<char, Vec<Member>, Error = Simple<char>> {
    return member_parser(element_parser)
        .separated_by(just(","))
        .collect();
}

fn member_parser(
    element_parser: impl Parser<char, Element, Error = Simple<char>>,
) -> impl Parser<char, Member, Error = Simple<char>> {
    return whitespace()
        .ignore_then(member_key_parser())
        .then_ignore(whitespace())
        .then_ignore(just(":"))
        .then(element_parser)
        .map(|(key, value)| Member { key, value });
}

fn member_key_parser() -> impl Parser<char, String, Error = Simple<char>> {
    return character_parser()
        .repeated()
        .at_least(1)
        .delimited_by(just('"'), just('"'))
        .map(|chars: Vec<char>| chars.into_iter().collect::<String>());
}

fn array_parser(
    element_parser: impl Parser<char, Element, Error = Simple<char>> + Clone,
) -> impl Parser<char, Value, Error = Simple<char>> {
    let empty = whitespace().delimited_by(just("["), just("]")).map(|_| {
        Value::Array(Array {
            elements: Vec::new(),
        })
    });

    let elements = elements_parser(element_parser)
        .delimited_by(just("["), just("]"))
        .map(|elements| Value::Array(Array { elements }));

    return empty.or(elements);
}

fn elements_parser(
    element_parser: impl Parser<char, Element, Error = Simple<char>> + Clone,
) -> impl Parser<char, Vec<Element>, Error = Simple<char>> {
    return element_parser.separated_by(just(",")).collect();
}

pub(crate) fn element_parser() -> impl Parser<char, Element, Error = Simple<char>> + Clone {
    return recursive(|element_parser| {
        whitespace()
            .ignore_then(value_parser(element_parser))
            .then_ignore(whitespace())
            .map(|value| Element { value })
    });
}

fn string_value_parser() -> impl Parser<char, Value, Error = Simple<char>> {
    return characters_parser()
        .delimited_by(just('"'), just('"'))
        .map(|val| Value::String(val));
}

pub(crate) fn characters_parser() -> impl Parser<char, CompositeString, Error = Simple<char>> {
    return variable_name_parser()
        .or(character_parser().map(|c| CompositeStringPart::Literal(c.to_string())))
        .repeated()
        .map(|parts| CompositeString { parts });
}

fn character_parser() -> impl Parser<char, char, Error = Simple<char>> {
    let valid_char = filter(|c: &char| {
        let in_range = *c >= '\u{0020}' && *c <= '\u{10FFF}';
        let is_special = *c == '"' || *c == '\\';
        in_range && !is_special
    })
    .map(|c| c);

    return escape_parser().or(valid_char);
}

fn escape_parser() -> impl Parser<char, char, Error = Simple<char>> {
    return just("\\").ignore_then(choice((
        just('"').to('"'),
        just("\\").to('\\'),
        just("/").to('/'),
        just("b").to('\u{0008}'), // Backspace
        just("f").to('\u{000C}'), // Form feed
        just("n").to('\n'),
        just("r").to('\r'),
        just("t").to('\t'),
        hex_parser(),
    )));
}

fn hex_parser() -> impl Parser<char, char, Error = Simple<char>> {
    return just("u")
        .ignore_then(one_of("0123456789abcdefABCDEF").repeated().exactly(4))
        .map(|digits: Vec<char>| {
            let as_string: String = digits.into_iter().collect();
            char::from_u32(u32::from_str_radix(&as_string, 16).unwrap()).unwrap()
        });
}

fn number_value_parser() -> impl Parser<char, Value, Error = Simple<char>> {
    return number_parser().map(|value| Value::Number(value));
}

fn number_parser() -> impl Parser<char, Number, Error = Simple<char>> {
    return exponent_parser().or(fraction_parser()).or(integer_parser());
}

fn integer_parser() -> impl Parser<char, Number, Error = Simple<char>> {
    let positive = int(10).map(|num: String| num.parse::<i64>().unwrap());
    let negative = just("-").ignore_then(positive).map(|num| -num);
    return positive.or(negative).map(|num| Number::Int(num));
}

fn fraction_parser() -> impl Parser<char, Number, Error = Simple<char>> {
    let positive = int(10)
        .then_ignore(just("."))
        .then(int(10))
        .map(|(whole, frac): (String, String)| format!("{whole}.{frac}").parse::<f64>().unwrap());
    let negative = just("-").ignore_then(positive).map(|num| -num);
    return positive.or(negative).map(|num| Number::Fraction(num));
}

fn exponent_parser() -> impl Parser<char, Number, Error = Simple<char>> {
    let sign = (one_of("+-").or_not()).map(|c: Option<char>| {
        if c.is_none() {
            "".to_string()
        } else {
            c.unwrap().to_string()
        }
    });
    let digits = one_of("0123456789")
        .repeated()
        .at_least(1)
        .map(|chars: Vec<char>| chars.into_iter().collect::<String>());

    return (sign.clone())
        .then(digits.clone())
        .then(just("."))
        .then(digits.clone())
        .then(one_of("eE"))
        .then(sign.clone())
        .then(digits.clone())
        .map(
            |((((((base_sign, base), dot), decimals), e_lit), exponent_sign), exponent)| {
                Number::Exponent(
                    [
                        base_sign,
                        base,
                        dot.to_string(),
                        decimals,
                        e_lit.to_string(),
                        exponent_sign.to_string(),
                        exponent,
                    ]
                    .concat(),
                )
            },
        );
}
