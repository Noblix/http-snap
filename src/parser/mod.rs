mod body_parser;
mod header_parser;
mod url_parser;
mod variable_parser;

use crate::types::*;
use chumsky::error::Simple;
use chumsky::Parser;

fn parser() -> impl Parser<char, HttpFile, Error = Simple<char>> {
    let base = variable_parser::variables_skipper()
        .ignore_then(url_parser::verb_parser())
        .then(url_parser::url_parser())
        .then(header_parser::headers_parser())
        .then(body_parser::body_parser())
        .map(|(((verb, url), headers), body)| HttpFile {
            verb,
            url,
            headers,
            body,
        });

    return base;
}

pub fn parse(input: &str) -> Result<HttpFile, Vec<Simple<char>>> {
    let without_variables = variable_parser::replace_variables(input);
    let result = parser().parse(without_variables);
    return result;
}
