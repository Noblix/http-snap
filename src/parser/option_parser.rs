use crate::types::SnapOptions;
use chumsky::error::Simple;
use chumsky::prelude::{just};
use chumsky::text::whitespace;
use chumsky::Parser;

pub(crate) fn options_parser() -> impl Parser<char, SnapOptions, Error = Simple<char>> {
    let repeated_spaces = just(' ').repeated();

    let no_options = whitespace().map(|_| SnapOptions {
        include_headers: true,
    });

    let include_headers = whitespace()
        .then(just("include-headers"))
        .then(repeated_spaces.clone())
        .then(just(":"))
        .then(repeated_spaces.clone())
        .ignore_then(just("true").or(just("false")))
        .map(|value| if value == "true" { true } else { false })
        .map(|value| SnapOptions {
            include_headers: value,
        });

    return include_headers.or(no_options);
}
