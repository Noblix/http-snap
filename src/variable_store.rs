use crate::types::{SnapResponse, Value};
use regex::Regex;
use std::collections::HashMap;

pub struct VariableStore {
    variables: HashMap<String, String>,
}

impl VariableStore {
    pub fn new() -> Self {
        return Self {
            variables: HashMap::new(),
        };
    }

    pub fn replace_variables(&mut self, text: &str, previous: &Option<SnapResponse>) -> String {
        let previous_headers_replaced = replace_previous_headers(text, &previous);
        let previous_body_replaced = replace_previous_body(&previous_headers_replaced, &previous);
        let new_variables = extract_variables(&previous_body_replaced);
        self.variables.extend(new_variables.into_iter());
        let text_without_variables = self.replace_local_variables(&previous_body_replaced);
        return text_without_variables;
    }

    fn replace_local_variables(&self, text: &str) -> String {
        let mut result: String = text.to_string();
        for (name, value) in &self.variables {
            let replace = ["{{", name, "}}"].concat();
            result = result.replace(&replace, value);
        }
        return result;
    }
}

fn replace_previous_body(text: &str, previous: &Option<SnapResponse>) -> String {
    let mut result: String = text.to_string();

    if let Some(previous) = previous {
        let body_regex = Regex::new(r#"\{\{previous\.body\.([^\n]+)\}\}"#).unwrap();
        let body_properties = body_regex.captures_iter(text).map(|c| c.extract());
        for (_, [path_to_property]) in body_properties {
            let parts: Vec<&str> = path_to_property.split(".").collect();
            let mut current_value = &previous.body.element.value;
            for part in parts {
                match current_value {
                    Value::Object(object) => {
                        current_value = if let Some(member) =
                            object.members.iter().find(|member| member.key == part)
                        {
                            &member.value.value
                        } else {
                            panic!("Unknown member")
                        }
                    }
                    _ => panic!("Unsupported field identification"),
                }
            }

            let replace = ["{{previous.body.", path_to_property, "}}"].concat();
            result = result.replace(&replace, &current_value.to_insertion_string(false));
        }
    }

    return result;
}

fn replace_previous_headers(text: &str, previous: &Option<SnapResponse>) -> String {
    let mut result: String = text.to_string();

    if let Some(previous) = previous {
        let header_regex =
            Regex::new(r#"\{\{previous\.headers\["([a-zA-Z0-9\-\_]+)"\]\}\}"#).unwrap();
        let header_names = header_regex.captures_iter(text).map(|c| c.extract());
        for (_, [header_name]) in header_names {
            let header = previous.headers.get(header_name);
            if let Some(value) = header {
                let replace = ["{{previous.headers[\"", header_name, "\"]}}"].concat();
                result = result.replace(&replace, value.to_str().unwrap().trim())
            }
        }
    }

    return result;
}

fn extract_variables(text: &str) -> HashMap<String, String> {
    let mut result = HashMap::new();

    for mut line in text.lines() {
        line = line.trim();
        if line.starts_with("@") {
            let parts: Vec<&str> = line.split("=").collect();
            if parts.len() >= 2 {
                let name = parts[0][1..].trim().to_string();
                let value = parts[1..].concat().trim().to_string();
                result.insert(name, value);
            }
        }

        let verb_options = ["GET", "DELETE", "PATCH", "POST", "PUT"];
        for prefix in verb_options {
            if line.starts_with(prefix) {
                break;
            }
        }
    }

    return result;
}
