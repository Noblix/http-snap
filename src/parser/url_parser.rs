use crate::parser::body_parser;
use crate::types::{CompositeString, HttpVerb};
use chumsky::error::Simple;
use chumsky::prelude::*;
use chumsky::Parser;

pub(crate) fn verb_parser() -> impl Parser<char, HttpVerb, Error = Simple<char>> {
    let verb = (just("GET").map(|_| HttpVerb::GET))
        .or(just("DELETE").map(|_| HttpVerb::DELETE))
        .or(just("PATCH").map(|_| HttpVerb::PATCH))
        .or(just("POST").map(|_| HttpVerb::POST))
        .or(just("PUT").map(|_| HttpVerb::PUT))
        .padded();

    return verb;
}

pub(crate) fn url_parser() -> impl Parser<char, CompositeString, Error = Simple<char>> {
    let url = body_parser::characters_parser();
    return url;
}
