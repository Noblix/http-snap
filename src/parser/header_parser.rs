use crate::parser::snapshot_parser::ignore_comparison_parser;
use crate::parser::variable_parser::variable_store_parser;
use crate::types::*;
use chumsky::error::Simple;
use chumsky::prelude::*;
use chumsky::text::whitespace;
use chumsky::Parser;
use std::ops::Add;

pub(crate) fn headers_parser(
    compare: bool,
) -> impl Parser<char, Vec<Header>, Error = Simple<char>> {
    if compare {
        return headers_compare_parser();
    }

    return headers_no_compare_parser();
}

fn headers_no_compare_parser() -> Box<dyn Parser<char, Vec<Header>, Error = Simple<char>>> {
    return Box::new(
        header_key()
            .then_ignore(just(':'))
            .then_ignore(repeated_spaces())
            .then(header_value())
            .then(variable_store_parser().or_not())
            .then_ignore(whitespace())
            .map(|((name, value), variable_store)| Header {
                name,
                value,
                variable_store,
                comparison: None,
            })
            .padded()
            .repeated(),
    );
}

fn headers_compare_parser() -> Box<dyn Parser<char, Vec<Header>, Error = Simple<char>>> {
    return Box::new(
        header_key()
            .then_ignore(just(':'))
            .then_ignore(repeated_spaces())
            .then(
                (ignore_comparison_parser().map(|_| (String::new(), Some(Comparison::Ignore))))
                    .or(header_value().map(|value| (value, Some(Comparison::Exact)))),
            )
            .then(variable_store_parser().or_not())
            .then_ignore(whitespace())
            .map(|((name, (value, comparison)), variable_store)| Header {
                name,
                value,
                variable_store,
                comparison,
            })
            .padded()
            .repeated(),
    );
}

fn repeated_spaces() -> impl Parser<char, Vec<char>, Error = Simple<char>> {
    return just(' ').repeated();
}

fn header_key() -> impl Parser<char, String, Error = Simple<char>> {
    return text::ident()
        .then(filter(|c: &char| c.is_ascii_alphanumeric() || c == &'-').repeated())
        .foldl(|start, c| start.add(&c.to_string()));
}

fn header_value() -> impl Parser<char, String, Error = Simple<char>> {
    return filter(|x: &char| x != &'\r' && x != &'\n' && x != &'{' && x != &'}')
        .repeated()
        .at_least(1)
        .map(|chars: Vec<char>| chars.into_iter().collect::<String>());
}
