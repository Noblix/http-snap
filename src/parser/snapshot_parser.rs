use crate::parser::body_parser::{body_parser, characters_parser};
use crate::parser::header_parser::headers_parser;
use crate::types::{Comparison, Element, Json, Number, Snapshot, Value};
use chumsky::error::Simple;
use chumsky::prelude::*;
use chumsky::text::whitespace;
use chumsky::Parser;

// Based on https://www.json.org/json-en.html
pub(crate) fn snapshot_parser() -> impl Parser<char, Snapshot, Error = Simple<char>> {
    let no_snapshot = end().map(|_| Snapshot {
        status: Number::Int(-1),
        headers: Vec::new(),
        body: Json {
            element: Element {
                value: Value::Null(),
                variable_store: None,
                comparison: Some(Comparison::Ignore),
            },
        },
    });

    let snapshot = just("SNAPSHOT")
        .ignore_then(whitespace())
        .ignore_then(status_parser())
        .then_ignore(whitespace())
        .then(headers_parser(true))
        .then_ignore(whitespace())
        .then(body_parser(true))
        .then_ignore(whitespace().then(end()))
        .map(|((status, headers), body)| Snapshot {
            status,
            headers,
            body,
        });

    return whitespace().ignore_then(no_snapshot.or(snapshot));
}

pub(crate) fn ignore_comparison_parser() -> impl Parser<char, Comparison, Error = Simple<char>> {
    return whitespace()
        .then(just("_"))
        .then_ignore(whitespace())
        .map(|_| Comparison::Ignore);
}

pub(crate) fn timestamp_format_parser() -> impl Parser<char, Comparison, Error = Simple<char>> {
    return whitespace()
        .then(just("timestamp"))
        .ignore_then(characters_parser().delimited_by(just("(\""), just("\")")))
        .map(|pattern| Comparison::TimestampFormat(pattern));
}

fn status_parser() -> impl Parser<char, Number, Error = Simple<char>> {
    return just("status: ")
        .ignore_then(one_of("0123456789").repeated().exactly(3))
        .map(|code: Vec<char>| {
            Number::Int(code
                .into_iter()
                .collect::<String>()
                .parse::<i64>()
                .unwrap())
        });
}