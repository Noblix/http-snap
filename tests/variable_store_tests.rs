use http_snap::parser::parse_file;
use http_snap::types::{
    Array, Element, Json, Member, Number, Object, SnapOptions, SnapResponse, Value,
};
use http_snap::variable_store::VariableStore;
use indoc::indoc;
use reqwest::header::{HeaderMap, HeaderName, HeaderValue};

#[test]
fn replace_previous_header_variable() {
    let previous = SnapResponse {
        options: SnapOptions {
            include_headers: true,
        },
        status: 200,
        headers: HeaderMap::from_iter([(
            HeaderName::from_static("next-page"),
            HeaderValue::from(2),
        )]),
        body: Json {
            element: Element {
                value: Value::Null(),
            },
        },
    };
    let input = indoc! {r#"
        GET https://localhost:5000/items?page={{previous.headers["next-page"]}}
        Accept: application/json
    "#};
    let mut sut = VariableStore::new();

    let result = sut.replace_variables(input, &Some(previous));

    let parsed = parse_file(&result).unwrap();
    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn replace_previous_body_variable_number() {
    let previous = SnapResponse {
        options: SnapOptions {
            include_headers: true,
        },
        status: 200,
        headers: HeaderMap::new(),
        body: Json {
            element: Element {
                value: Value::Object(Object {
                    members: vec![Member {
                        key: "id".to_string(),
                        value: Element {
                            value: Value::Number(Number::Int(123)),
                        },
                    }],
                }),
            },
        },
    };
    let input = indoc! {r#"
        GET https://localhost:5000/items/{{previous.body.id}}
        Accept: application/json
    "#};
    let mut sut = VariableStore::new();

    let result = sut.replace_variables(input, &Some(previous));

    let parsed = parse_file(&result).unwrap();
    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn replace_previous_body_variable_string() {
    let previous = SnapResponse {
        options: SnapOptions {
            include_headers: true,
        },
        status: 200,
        headers: HeaderMap::new(),
        body: Json {
            element: Element {
                value: Value::Object(Object {
                    members: vec![Member {
                        key: "id".to_string(),
                        value: Element {
                            value: Value::String(
                                "97b9b76e-dff0-4400-a974-b59fde32724c".to_string(),
                            ),
                        },
                    }],
                }),
            },
        },
    };
    let input = indoc! {r#"
        GET https://localhost:5000/items/{{previous.body.id}}
        Accept: application/json
    "#};
    let mut sut = VariableStore::new();

    let result = sut.replace_variables(input, &Some(previous));

    let parsed = parse_file(&result).unwrap();
    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn replace_previous_body_variable_bool() {
    let previous = SnapResponse {
        options: SnapOptions {
            include_headers: true,
        },
        status: 200,
        headers: HeaderMap::new(),
        body: Json {
            element: Element {
                value: Value::Object(Object {
                    members: vec![Member {
                        key: "is_anonymous".to_string(),
                        value: Element {
                            value: Value::Boolean(true),
                        },
                    }],
                }),
            },
        },
    };
    let input = indoc! {r#"
        GET https://localhost:5000/items
        Accept: application/json
        
        {
            "anonymous": {{previous.body.is_anonymous}}
        }
    "#};
    let mut sut = VariableStore::new();

    let result = sut.replace_variables(input, &Some(previous));

    let parsed = parse_file(&result).unwrap();
    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn replace_previous_body_variable_null() {
    let previous = SnapResponse {
        options: SnapOptions {
            include_headers: true,
        },
        status: 200,
        headers: HeaderMap::new(),
        body: Json {
            element: Element {
                value: Value::Object(Object {
                    members: vec![Member {
                        key: "filters".to_string(),
                        value: Element {
                            value: Value::Null(),
                        },
                    }],
                }),
            },
        },
    };
    let input = indoc! {r#"
        GET https://localhost:5000/items
        Accept: application/json
        
        {
            "filters": {{previous.body.filters}}
        }
    "#};
    let mut sut = VariableStore::new();

    let result = sut.replace_variables(input, &Some(previous));

    let parsed = parse_file(&result).unwrap();
    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn replace_previous_body_variable_object() {
    let previous = SnapResponse {
        options: SnapOptions {
            include_headers: true,
        },
        status: 200,
        headers: HeaderMap::new(),
        body: Json {
            element: Element {
                value: Value::Object(Object {
                    members: vec![Member {
                        key: "nested".to_string(),
                        value: Element {
                            value: Value::Object(Object {
                                members: vec![
                                    Member {
                                        key: "inside_string".to_string(),
                                        value: Element {
                                            value: Value::String("should be in quotes".to_string()),
                                        },
                                    },
                                    Member {
                                        key: "inside_number".to_string(),
                                        value: Element {
                                            value: Value::Number(Number::Int(82)),
                                        },
                                    },
                                ],
                            }),
                        },
                    }],
                }),
            },
        },
    };
    let input = indoc! {r#"
        GET https://localhost:5000/items
        Accept: application/json
        
        {
            "nested": {{previous.body.nested}}
        }
    "#};
    let mut sut = VariableStore::new();

    let result = sut.replace_variables(input, &Some(previous));

    let parsed = parse_file(&result).unwrap();
    insta::assert_debug_snapshot!(parsed);
}

#[test]
fn replace_previous_body_variable_array() {
    let previous = SnapResponse {
        options: SnapOptions {
            include_headers: true,
        },
        status: 200,
        headers: HeaderMap::new(),
        body: Json {
            element: Element {
                value: Value::Object(Object {
                    members: vec![Member {
                        key: "mixed_array".to_string(),
                        value: Element {
                            value: Value::Array(Array {
                                elements: vec![
                                    Element {
                                        value: Value::Object(Object {
                                            members: vec![Member {
                                                key: "inside".to_string(),
                                                value: Element {
                                                    value: Value::String(
                                                        "should be in quotes".to_string(),
                                                    ),
                                                },
                                            }],
                                        }),
                                    },
                                    Element {
                                        value: Value::String("Hello".to_string()),
                                    },
                                    Element {
                                        value: Value::Number(Number::Int(42)),
                                    },
                                ],
                            }),
                        },
                    }],
                }),
            },
        },
    };
    let input = indoc! {r#"
        GET https://localhost:5000/items
        Accept: application/json
        
        {
            "array": {{previous.body.mixed_array}}
        }
    "#};
    let mut sut = VariableStore::new();

    let result = sut.replace_variables(input, &Some(previous));

    let parsed = parse_file(&result).unwrap();
    insta::assert_debug_snapshot!(parsed);
}
