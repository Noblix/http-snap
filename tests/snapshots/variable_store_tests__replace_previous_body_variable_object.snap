---
source: tests/variable_store_tests.rs
assertion_line: 192
expression: parsed
---
HttpFile {
    options: SnapOptions {
        include_headers: true,
    },
    verb: GET,
    url: "https://localhost:5000/items",
    headers: [
        Header {
            name: "Accept",
            value: "application/json",
        },
    ],
    body: Json {
        element: Element {
            value: Object(
                Object {
                    members: [
                        Member {
                            key: "nested",
                            value: Element {
                                value: Object(
                                    Object {
                                        members: [
                                            Member {
                                                key: "inside_string",
                                                value: Element {
                                                    value: String(
                                                        "should be in quotes",
                                                    ),
                                                },
                                            },
                                            Member {
                                                key: "inside_number",
                                                value: Element {
                                                    value: Number(
                                                        Int(
                                                            82,
                                                        ),
                                                    ),
                                                },
                                            },
                                        ],
                                    },
                                ),
                            },
                        },
                    ],
                },
            ),
        },
    },
    snapshot: Snapshot {
        status: Exact(
            Null,
        ),
        headers: [],
        body: JsonComparer {
            element: ElementComparer {
                value: Exact(
                    Null,
                ),
            },
        },
    },
}
